Various Updates and Bugfixes

PeterDinda opened the pull request:
Major new things here are the addition of futures, and a refactor of the openmp code to make room for more than one openmp implementation.

Bug fixes:   timer race, incorrect handling of stack checks in scheduler (only visible with fibers, which is n upcoming addition), use of low-level memcpy/memset in vga in order to be compatible with KVM (this is a more general issue - compiler may use SSE to talk to MMIO, which breaks in KVM and similar because they don't emulate SSE for MMIO).  

Also has minor restructuring of task code.

PeterDinda committed:
Restructuring of current OpenMP code to make room for CLANG/LLVM version

PeterDinda committed:
Fix minor issues (shadow vars) in vc and waitqueue (not bugs)

PeterDinda committed:
Minor cleanup of task code

PeterDinda committed:
MTRR tweak for error output

PeterDinda committed:
Add futures (initial implementation)

This adds futures to the kernel-level abstractions

Futures can be waited on using spinning or blocking
Currently, a wait queue is involved regardless.

Commit:
Bug fix - truly deactivate stack checks in scheduler when asked to do so

MichaelCuevas committed:
Bug fix - fix timer race condition

PeterDinda committed:
Add a low-level memcpy routine, and use it in the VGA code.

Reason:  clang will implement memcpy using SSE.   So now there
is SSE code touching MMIO space.  KVM (and presumably other VMMs)
don't emulate SSE access to MMIO space.

PeterDinda committed:
Update default and debug configs

khale commented:
This pull request **introduces 2 alerts** and **fixes 2** when merging 01a4950fe27be0f30d27031fcbd400b9622e08c5 into d52494940a96704b0cb3ae6964a12db36361d3e0 - [view on LGTM.com](https://lgtm.com/projects/g/HExSA-Lab/nautilus/rev/pr-1d6fed03a3821981ad57952ae7f5ef6994f52fa0)

**new alerts:**

* 1 for Wrong type of arguments to formatting function
* 1 for Function declared in block

**fixed alerts:**

* 2 for Local variable hides global variable

---

**Warning** - Automated code review for HExSA-Lab/nautilus will be disabled on October 1, 2019. You can avoid this by <a href="https://github.com/apps/lgtm-com/installations/new">installing the LGTM.com GitHub App</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

PeterDinda committed:
Minor warning fix in futures

khale commented:
This pull request **fixes 2 alerts** when merging 77fa38e84276dc69ca9418416ecec35dd1f51516 into d52494940a96704b0cb3ae6964a12db36361d3e0 - [view on LGTM.com](https://lgtm.com/projects/g/HExSA-Lab/nautilus/rev/pr-3b612eeed940d57a455727dbff922c66e0e42cfd)

**fixed alerts:**

* 2 for Local variable hides global variable

---

**Warning** - Automated code review for HExSA-Lab/nautilus will be disabled on October 1, 2019. You can avoid this by <a href="https://github.com/apps/lgtm-com/installations/new">installing the LGTM.com GitHub App</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*

PeterDinda committed:
Do all VGA accesses using movb and movw instructions.

1. This is needed for VMs (e.g. KVM) that do not emulate SSE
2. This makes VGA accessible before SSE is enabled

This patch also moves vga init very early in the boot process so
we can get output right away.

PeterDinda committed:
Fix gcc compile issue

gcc versions 6 and above use position independence by default
this breaks us badly, especialy since with mcmodel=large, gcc
does not seem to get this correct in many cases.

(found by Mike Cuevas)

PeterDinda committed:
Minor mbutils bug fix

check first, then iterate...

khale commented:
Looks good

khale commented:
This pull request **fixes 2 alerts** when merging 3068a7559a0b820ab06ed401a2a945da8d100edf into d52494940a96704b0cb3ae6964a12db36361d3e0 - [view on LGTM.com](https://lgtm.com/projects/g/HExSA-Lab/nautilus/rev/pr-258797df6aade128c0b485afdb27e6e935028dac)

**fixed alerts:**

* 2 for Local variable hides global variable

---

**Warning** - Automated code review for HExSA-Lab/nautilus will be disabled on October 1, 2019. You can avoid this by <a href="https://github.com/apps/lgtm-com/installations/new">installing the LGTM.com GitHub App</a>. Read about the <a href="https://lgtm.com/blog/announcing_github_apps_migration">benefits of migrating to GitHub Apps</a> in the blog.

---

*Comment posted by [LGTM.com](https://lgtm.com)*
